### PR TITLE
[FLINK-4116] Metrics documentation

### PR DESCRIPTION
**Pull request by zentol**

### CODE REVIEW

File: docs/apis/common/index.md [hunk 1]
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
 the execution plan before executing the Flink job.
 
 {% top %}
+
+Metrics
+-------------------
+
+Flink exposes a metric system that allows users gather and expose metrics to external systems.
```
Comment (tillrohrmann) [inline docs/apis/common/index.md]: "allows users to gather and ..." or "allows user gathering and ..."

File: docs/apis/common/index.md [hunk 2]
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+
+### Registering metrics
+
+You can access the metric system from any user function that extends [RichFunction]({{ site.baseurl }}/apis/common/index.html#rich-functions) by calling getRuntimeContext().getMetricGroup().
```
Comment (tillrohrmann) [inline docs/apis/common/index.md]: `getRuntimeContext().getMetricGroup` should be enclosed by back ticks.

File: docs/apis/common/index.md [hunk 3]
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+This method returns a MetricGroup object on which you can create and register new metrics.
```
Comment (tillrohrmann) [inline docs/apis/common/index.md]: `MetricGroup` should have backticks.

File: docs/apis/common/index.md [hunk 4]
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+
+If you want to count the number of records your user function has received you could use a Counter like this:
```
Comment (tillrohrmann) [inline docs/apis/common/index.md]: `Counter` in backticks.

File: docs/apis/common/index.md [hunk 5]
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+  }
+
+  @public Integer map(String value) throws Exception {
+    // increment counter
+    this.counter.inc();
+    ...
+  }
+}
+
+{% endhighlight %}
```
Comment (tillrohrmann) [inline docs/apis/common/index.md]: Maybe we should also demonstrate how to register a custom metric.

File: docs/apis/common/index.md [hunk 6]
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+
+### Metric types
+
+Flink supports Counters, Gauges and Histograms.
```
Comment (tillrohrmann) [inline docs/apis/common/index.md]: backticks or writing counters, gauges and histograms without a capital letter.

File: docs/apis/common/index.md [hunk 7]
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+
+A Counter is used to count something. The current value can be in- or decremented using `inc([long n])` or `dec([long n])`.
```
Comment (tillrohrmann) [inline docs/apis/common/index.md]: I've never seen the notation `inc([long n])` in the Flink documentation. Maybe it's better to specify the parameterless and the parameter variant explicitly.

File: docs/apis/common/index.md [hunk 8]
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+You can create and registers a Counter by calling `counter(String name)` on a MetricGroup. Alternatively, you can pass an instance of your own Counter implementation along with the name.
```
Comment (tillrohrmann) [inline docs/apis/common/index.md]: register without *s*

File: docs/apis/common/index.md [hunk 9]
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+
+A Gauge provides a value of any type on demand. In order to use a Gauge you must first create a class that implements the `org.apache.flink.metrics.Gauge` interface.
+There is not restriction for the type of the returned value.
+You can register a gauge by calling `gauge(String name, Gauge gauge)` on a MetricGroup.
+
+A Histogram measure the distribution of long values.
+You can register one by calling histogram(String name, Histogram histogram) on a MetricGroup.
+
+Flink only provides an interface for Histograms, but offers a Wrapper that allows usage of Codahale/DropWizard Histograms. (org.apache.flink.dropwizard.metrics.DropWizardHistogramWrapper)
```
Comment (tillrohrmann) [inline docs/apis/common/index.md]: I think it would be better to present the different metric types either as a descriptive list similar to the description environment of latex or by using sub headings. It could also be more descriptive to give for each metric type a short code snippet how to register and interacting with it. That would explain their interface which is important for the user.

Comment (tillrohrmann) [inline docs/apis/common/index.md]: Maybe we could also say which module contains the `DropWizardHistogramWrapper`

File: docs/apis/common/index.md [hunk 10]
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+
+### Scope
+
+Every registered metric has an automatically assigned scope which represent the entities it is tied to. By default a metric that is registered in a user function will be scoped to the Operator in which the function runs, the Task/Job it belongs to and the TaskManager/Host it is executed on. This is referred to as the "system scope".
```
Comment (tillrohrmann) [inline docs/apis/common/index.md]: which represents

Comment (tillrohrmann) [inline docs/apis/common/index.md]: operator with small letter.

Comment (tillrohrmann) [inline docs/apis/common/index.md]: task/job with small letter

Comment (tillrohrmann) [inline docs/apis/common/index.md]: Same for task manager/host

File: docs/apis/common/index.md [hunk 11]
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+
+You can define an additonal "user scope" by calling the MetricGroup#addGroup((int/String) name) method.
```
Comment (tillrohrmann) [inline docs/apis/common/index.md]: explicit enumeration of the different `addGroup` methods is better imo.

File: docs/apis/common/index.md [hunk 12]
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+
+{% highlight java %}
+
+counter2 = getRuntimeContext().getMetricGroup().addGroup("MyMetrics").counter("myCounter2");
+
+{% endhighlight %}
+
+The name under which a metric is exported is based on both scopes and the name passed in the `counter() call. The order is always \<system_scope>\<user_scope>\<name>.
```
Comment (tillrohrmann) [inline docs/apis/common/index.md]: counter has only one backtick.

Comment (tillrohrmann) [inline docs/apis/common/index.md]: Don't we insert a scope separating character in between the system, user and name scope?

File: docs/apis/common/index.md [hunk 13]
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+
+The system scope allows the reported name to contain contextual information like the name of job it was registered in without requiring the user to pass this information manually.
+
+How the system scope affects the reported name for a metric can be modified by setting the following keys in the flink-conf.yaml. 
+Each of these keys expect a format string that may contain constants (e.g. "taskmanager") and variables (e.g. "\<task_id>") which will be replaced at runtime.
```
Comment (tillrohrmann) [inline docs/apis/common/index.md]: The config key definition should also be mirrored to https://ci.apache.org/projects/flink/flink-docs-master/setup/config.html. Actually I think that the config keys should be moved from `MetricRegistry` to `ConfigConstants`.

File: docs/apis/common/index.md [hunk 14]
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+- JobManager: \<host>
+- TaskManager: \<host>, \<tm_id>
+- Job: \<job_id>, \<job_name>
+- Task: \<task_id>, \<task_name>, \<task_attempt_id>, \<task_attempt_num>, \<subtask_index>
+- Operator: \<operator_name>, \<subtask_index>
+
+There is no restriction on the order of variables.
+
+Which format is applied to a metric depends on which entities it was scoped to.
+The `metrics.scope.operator` format will be applied to all metrics that were scoped to an Operator.
```
Comment (tillrohrmann) [inline docs/apis/common/index.md]: operator with small letter

File: docs/apis/common/index.md [hunk 15]
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+The `metrics.scope.tm.job` format will be applied to all metrics that were only scoped to a Job and a TaskManager.
```
Comment (tillrohrmann) [inline docs/apis/common/index.md]: job and task manager.

Comment (tillrohrmann) [inline docs/apis/common/index.md]: Maybe it would be good to add the explanation of the different scopes directly to the enumeration of the config keys instead of listing them in this paragraph.

File: docs/apis/common/index.md [hunk 16]
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+
+This allows you to define explicitly how metrics on every level should be reported.
```
Comment (tillrohrmann) [inline docs/apis/common/index.md]: Giving a small example could be helpful.

File: docs/apis/common/index.md [hunk 17]
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+Metrics can be exposed to an external system by configuring a reporter in the `conf/flink-conf.yaml`.
+
+- `metrics.reporter.class`: The class of the reporter to use.
+  - Example: org.apache.flink.metrics.reporter.JMXReporter)
+- `metrics.reporter.arguments`: A list of named parameters that are passed to the reporter.
+  - Example: --host localhost --port 9010)
+- `metrics.reporter.interval`: The interval between reports.
+  - Example: 10 SECONDS)
+
+By default Flink uses JMX to expose metrics.
```
Comment (tillrohrmann) [inline docs/apis/common/index.md]: JMX should also get a sub heading.

File: docs/apis/common/index.md [hunk 18]
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+
+The port for JMX can be configured by setting the `metrics.jmx.port` key. This parameter expects either a single port
+or a port range, with the default being 9010-9025. Which port is used in the end is shown in the relevant Job-/TaskManager log.
+
+All non-JMXReporters are not part of the distribution and have to be added to the classpath manually. (usually by putting the jar into /lib)
```
Comment (tillrohrmann) [inline docs/apis/common/index.md]: parenthesis starts after the period

File: docs/apis/common/index.md [hunk 19]
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+
+You can write your own Reporter by implementing the `org.apache.flink.metrics.reporter.MetricReporter` interface.
+If the Reporter should send out reports regularly you have to implement the Scheduled interface as well.
```
Comment (tillrohrmann) [inline docs/apis/common/index.md]: Reporter and Scheduled should be put into backticks.

Comment (tillrohrmann) [inline docs/apis/common/index.md]: Maybe we should explain what `Scheduled` indicates.

File: docs/apis/common/index.md [hunk 20]
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+
+Flink supports the following systems with their respective configuration parameters:
+
+#### Ganglia (org.apcahe.flink.metrics.ganglia.GangliaReporter)
+- `host`
+- `port`
+- `dmax`
+- `tmax`
+- `ttl`
+- `addressingMode` (UNICAST/MULTICAST)
```
Comment (tillrohrmann) [inline docs/apis/common/index.md]: Explanation for all of the parameters required.

Comment (tillrohrmann) [inline docs/apis/common/index.md]: Which flink module do I have to add for this reporter? Maybe give code snippet which you can easily add to your `pom.xml` file.

File: docs/apis/common/index.md [hunk 21]
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+
+#### Graphite (org.apache.flink.metrics.graphite.GraphiteReporter)
+- `host`
+- `port`
```
Comment (tillrohrmann) [inline docs/apis/common/index.md]: Explanation of parameters

File: docs/apis/common/index.md [hunk 22]
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+
+#### StatsD (org.apache.flink.metrics.statsd.StatsDReporter)
+- `host`
+- `port`
```
Comment (tillrohrmann) [inline docs/apis/common/index.md]: Explanation of parameters

File: docs/apis/common/index.md [hunk 23]
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+    <tr>
+      <td><strong>Operator</strong></td>
+      <td>
+        <p>numRecordsIn</p>
+        <p>numRecordsOut</p>
+        <p>numSplitsProcessed</p>
+      </td>
+    </tr>
+  </tbody>
+</table>
```
Comment (tillrohrmann) [inline docs/apis/common/index.md]: Explanation of metrics could be helpful.

File: docs/apis/common/index.md [hunk 24]
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+- `addressingMode` (UNICAST/MULTICAST)
+
+#### Graphite (org.apache.flink.metrics.graphite.GraphiteReporter)
+- `host`
+- `port`
+
+#### StatsD (org.apache.flink.metrics.statsd.StatsDReporter)
+- `host`
+- `port`
+
```
Comment (tillrohrmann) [inline docs/apis/common/index.md]: We should also say how the different reporters expose certain metrics. E.g. the histogram metric reports the mean, median, percentiles etc. The names under which this happens might differ between different reporters.

**Comment by tillrohrmann** [comment]
Good first draft of the metrics documentation @zentol. I think the documentation should be a bit more verbose containing more explanations and code examples. This will help the user to quickly get started and understanding the metrics system.

File: docs/apis/common/index.md (hunk 19 above)
Comment (zentol) [inline docs/apis/common/index.md]: Doesn't 'If the Reporter should send out reports regularly...' indicate what `Scheduled` means?

File: docs/apis/common/index.md (hunk 12 above)
Comment (zentol) [inline docs/apis/common/index.md]: kind of. i guess technically the delimiter is part of every group. I wanted to avoid adding `.` between the scope parts as it may lead to the misunderstanding that there are always only 2 dots in the name.

Comment (tillrohrmann) [inline docs/apis/common/index.md]: We could put it into backticks.

File: docs/apis/common/index.md [hunk 25]
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+or a port range, with the default being 9010-9025. Which port is used in the end is shown in the relevant Job-/TaskManager log.
+
+All non-JMXReporters are not part of the distribution and have to be added to the classpath manually, usually by putting the jar into /lib.
+
+You can write your own Reporter by implementing the `org.apache.flink.metrics.reporter.MetricReporter` interface.
+If the Reporter should send out reports regularly you have to implement the Scheduled interface as well.
+
+Flink supports the following systems with their respective configuration parameters:
+
+#### Ganglia (org.apcahe.flink.metrics.ganglia.GangliaReporter)
```
Comment (rmetzger) [inline docs/apis/common/index.md]: typo in apcahe

File: docs/apis/common/index.md (hunk 13 above)
Comment (rmetzger) [inline docs/apis/common/index.md]: I agree that we should mirror the config keys to the configuration page as well + move the config keys.

File: docs/apis/common/index.md (hunk 12 above)
Comment (zentol) [inline docs/apis/common/index.md]: I don't understand how backticks solve the issue i mentioned.

Comment (tillrohrmann) [inline docs/apis/common/index.md]: Hmm I think you're right, since the periods won't probably be rendered differently. But you could put the whole name in quotation marks. Then it should be clear.

**Comment by zentol** [comment]
I've addressed all comments, except the config page.

**Comment by zentol** [comment]
also missing: specific reporter naming

File: docs/apis/common/index.md [hunk 26]
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+  }
+}
+{% endhighlight %}
+
+Flink only provides an interface for Histograms, but offers a Wrapper that allows usage of Codahale/DropWizard Histograms. (org.apache.flink.dropwizard.metrics.DropWizardHistogramWrapper)
+This wrapper is contained in the `flink-metrics-dropwizard` module.
+
+### Scope
+
+Every registered metric has an automatically assigned scope which represents the entities it is tied to. By default a metric that is registered in a user function will be scoped to the operator in which the function runs, the task/job it belongs to and the taskManager/host it is executed on. This is referred to as the "system scope".
```
Comment (tillrohrmann) [inline docs/apis/common/index.md]: taskManager -> task manager

File: docs/apis/common/index.md [hunk 27]
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+
+You can define an additonal "user scope" by calling the either `MetricGroup#addGroup(String name)` or `MetricGroup#addGroup(int name)`.
```
Comment (tillrohrmann) [inline docs/apis/common/index.md]: "the" before "either" can be removed.

File: docs/apis/common/index.md [hunk 28]
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+This hierarchy also defines which variables may be accessed. The `tm.operator` format may contain variables for jobs, whereas `tm.job` may not contain
+variables for operators. There is no restriction on order of variables.
+
+There is no restriction on the order of variables.
+
+The default scope for operator metrics will result in a metric name akin to `localhost.taskmanager.1234.MyJob.MyOperator.0`
+
+If you also want to include the task name, but omit the taskmanager information you can specify the following format:
+
+`metrics.scope.tm.operator: \<host>.\<job_name>.\<task_name>.\<operator_name>.\<subtask_index>`
```
Comment (tillrohrmann) [inline docs/apis/common/index.md]: Do we need the `\` in front of the variables? They are also shown in the rendered html document.

File: docs/apis/common/index.md [hunk 29]
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+Metrics can be exposed to an external system by configuring a reporter in the `conf/flink-conf.yaml`.
+
+- `metrics.reporter.class`: The class of the reporter to use.
+  - Example: org.apache.flink.metrics.reporter.JMXReporter
+- `metrics.reporter.arguments`: A list of named parameters that are passed to the reporter.
+  - Example: --host localhost --port 9010
+- `metrics.reporter.interval`: The interval between reports.
+  - Example: 10 SECONDS
+
+You can write your own Reporter by implementing the `org.apache.flink.metrics.reporter.MetricReporter` interface.
```
Comment (tillrohrmann) [inline docs/apis/common/index.md]: Either putting `Reporter` in backticks or using a small letter would be better.

File: docs/apis/common/index.md [hunk 30]
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+If the Reporter should send out reports regularly you have to implement the Scheduled interface as well.
```
Comment (tillrohrmann) [inline docs/apis/common/index.md]: Same applies to `Scheduled`

**Comment by tillrohrmann** [comment]
The documentation has reached a really good state. I had only some minor comments left.

File: docs/apis/common/index.md (hunk 28 above)
Comment (zentol) [inline docs/apis/common/index.md]: Without the `\` they are interpreted as tags and not as text.

I used the build_docs.sh script to preview the docs, and the backslashes did not appear.

But I could (and probably should) replace all the < and > with `&lt;` and `&gt;` respectively.

Comment (tillrohrmann) [inline docs/apis/common/index.md]: I used Chrome to look at the docs. Maybe it depends on the browser. Replacing them by `&lt;` and `&gt;` sounds like the right way to go.

File: docs/apis/common/index.md (hunk 13 above)
Comment (uce) [inline docs/apis/common/index.md]: I agree, but note that there are other components who also do it outside of `ConfigConstants`.

File: docs/apis/common/index.md [hunk 31]
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
 the execution plan before executing the Flink job.
 
 {% top %}
+
+Metrics
```
Comment (uce) [inline docs/apis/common/index.md]: I would make this a separate page under `Programming Guide`. It's guide hidden under the `Basic API` concepts although I understand that it it is technically the right place for it.

File: docs/apis/common/index.md [hunk 32]
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+This method returns a `MetricGroup` object on which you can create and register new metrics.
+
+### Metric types
+
+Flink supports `Counters`, `Gauges` and `Histograms`.
+
+#### Counter
+
+A `Counter` is used to count something. The current value can be in- or decremented using `inc()/inc(long n)` or `dec()/dec(long n)`.
+You can create and register a `Counter` by calling `counter(String name)` on a MetricGroup.
```
Comment (uce) [inline docs/apis/common/index.md]: Orthogonal discussion, but the name `counter` to create a counter (and others respectively) is a style break imo compared the other classes which follow Java-bean style. Is that something we should change for the release? I'm definitely fine with keeping it the way it is, but it's something I've noticed.

File: docs/apis/common/index.md [hunk 33]
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+
+{% highlight java %}
+
+public class MyMapper extends RichMapFunction<String, Integer> {
+  ...
+
+  @Override
+  public void open(Configuration config) {
+    // register a custom counter
+    this.counter = getRuntimeContext().getmetricGroup().counter("myCustomCounter", new CustomCounter());
```
Comment (uce) [inline docs/apis/common/index.md]: the line breaks in my browser are bad, maybe move everything after `getMetricGroup()` to the next line? Furthermore, `getMetricGroup` has a lower letter `m`.

File: docs/apis/common/index.md [hunk 34]
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+  }
+  ...
+}
+
+{% endhighlight %}
+
+#### Gauge
+
+A `Gauge` provides a value of any type on demand. In order to use a `Gauge` you must first create a class that implements the `org.apache.flink.metrics.Gauge` interface.
+There is not restriction for the type of the returned value.
```
Comment (uce) [inline docs/apis/common/index.md]: btw does it not need to be serializable?

File: docs/apis/common/index.md [hunk 35]
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+
+{% highlight java %}
+
+public class MyMapper extends RichMapFunction<String, Integer> {
+  private int valueToExpose;
+
+  @Override
+  public void open(Configuration config) {
+    // register the gauge
+    getRuntimeContext().getmetricGroup().gauge("MyGauge", new Gauge<Integer>() {
```
Comment (uce) [inline docs/apis/common/index.md]: `getmetricGroup` with lower letter `m`

File: docs/apis/common/index.md [hunk 36]
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+      @Override
+      public Integer getValue() {
+        return valueToExpose;
+      }});
+    ...
```
Comment (uce) [inline docs/apis/common/index.md]: My personal taste, but I think it's better to just remove this (and other) three dots from the examples. Makes copy pasting easier and it should be clear from context that other code can be added as needed.

File: docs/apis/common/index.md [hunk 37]
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+  }
+  ...
+}
+
+{% endhighlight %}
+
+#### Histogram
+
+A Histogram measure the distribution of long values.
```
Comment (uce) [inline docs/apis/common/index.md]: measureS with s

File: docs/apis/common/index.md (hunk 32 above)
Comment (zentol) [inline docs/apis/common/index.md]: I'm not too happy with the names either.

File: docs/apis/common/index.md [hunk 38]
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+You can register one by calling histogram(String name, Histogram histogram) on a MetricGroup.
+
+{% highlight java %}
+public class MyMapper extends RichMapFunction<Long, Integer> {
+  private Histogram histogram;
+
+  @Override
+  public void open(Configuration config) {
+    // create and register a counter
+    this.histogram = getRuntimeContext().getMetricGroup().histogram("myHistogram", new MyHistogram());
```
Comment (uce) [inline docs/apis/common/index.md]: same line break issue as the custom counter

File: docs/apis/common/index.md (hunk 34 above)
Comment (zentol) [inline docs/apis/common/index.md]: no, reporters generally use toString() on the value returned. that should probably be documented though.

File: docs/apis/common/index.md [hunk 39]
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+  }
+  ...
+}
+
+{% endhighlight %}
+
+#### Histogram
+
+A Histogram measure the distribution of long values.
+You can register one by calling histogram(String name, Histogram histogram) on a MetricGroup.
```
Comment (uce) [inline docs/apis/common/index.md]: missing code formating for `histogram(...)` (via wrap in backtick)

File: docs/apis/common/index.md [hunk 40]
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+  }
+
+  @public Integer map(Long value) throws Exception {
+    this.histogram.update(value);
+    ...
+  }
+}
+{% endhighlight %}
+
+Flink only provides an interface for Histograms, but offers a Wrapper that allows usage of Codahale/DropWizard Histograms. (org.apache.flink.dropwizard.metrics.DropWizardHistogramWrapper)
```
Comment (uce) [inline docs/apis/common/index.md]: I would rephrase this along the lines of "Flink does not provide a default implementation for `Histogram`, but offers a wrapper that allows usage of Dropwizard histograms (see LINK TO CASS ON GITHUB).

File: docs/apis/common/index.md [hunk 41]
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+This wrapper is contained in the `flink-metrics-dropwizard` module.
```
Comment (uce) [inline docs/apis/common/index.md]: - Link to how a user can add this module
- I would also add an example for this actually (as everyone who wants to use Histogram will need this)

File: docs/apis/common/index.md (hunk 28 above)
Comment (uce) [inline docs/apis/common/index.md]: Same for me in Safari

File: docs/apis/common/index.md [hunk 42]
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+
+- JobManager: \<host>
+- TaskManager: \<host>, \<tm_id>
+- Job: \<job_id>, \<job_name>
+- Task: \<task_id>, \<task_name>, \<task_attempt_id>, \<task_attempt_num>, \<subtask_index>
+- Operator: \<operator_name>, \<subtask_index>
+
+### Reporter
+
+Metrics can be exposed to an external system by configuring a reporter in the `conf/flink-conf.yaml`.
```
Comment (uce) [inline docs/apis/common/index.md]: Either `in conf/flink-conf.yaml` or `in the ... file`

File: docs/apis/common/index.md [hunk 43]
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+
+- `metrics.reporter.class`: The class of the reporter to use.
+  - Example: org.apache.flink.metrics.reporter.JMXReporter
+- `metrics.reporter.arguments`: A list of named parameters that are passed to the reporter.
+  - Example: --host localhost --port 9010
+- `metrics.reporter.interval`: The interval between reports.
+  - Example: 10 SECONDS
+
+You can write your own `Reporter` by implementing the `org.apache.flink.metrics.reporter.MetricReporter` interface.
+If the Reporter should send out reports regularly you have to implement the `Scheduled` interface as well.
```
Comment (uce) [inline docs/apis/common/index.md]: Are there any reporters which are not scheduled?

File: docs/apis/common/index.md [hunk 44]
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+
+By default Flink uses JMX to expose metrics.
+All non-JMXReporters are not part of the distribution and have to be added to the classpath manually, either by putting the jar into /lib
```
Comment (uce) [inline docs/apis/common/index.md]: I think we also link to an explanation in another docs page for this

File: docs/apis/common/index.md [hunk 45]
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+or including it in the job jar.
+
+Flink supports the following systems:
```
Comment (uce) [inline docs/apis/common/index.md]: I think the `:` and the following sub sections don't go well together. I think either removing the sentence or saying something like `The following sections list the supported reporters` is better

File: docs/apis/common/index.md [hunk 46]
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+The port for JMX can be configured by setting the `metrics.jmx.port` key. This parameter expects either a single port
+or a port range, with the default being 9010-9025. The used port is shown in the relevant Job-/TaskManager log.
+
+#### Ganglia (org.apache.flink.metrics.ganglia.GangliaReporter)
+Dependency:
+{% highlight xml %}
+<dependency>
+      <groupId>org.apache.flink</groupId>
+      <artifactId>flink-metrics-ganglia</artifactId>
+      <version>1.1-SNAPSHOT</version>
```
Comment (uce) [inline docs/apis/common/index.md]: We need to use version variables instead of hard coded values here (and in other places)

File: docs/apis/common/index.md [hunk 47]
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+
+The default scope for operator metrics will result in a metric name akin to `localhost.taskmanager.1234.MyJob.MyOperator.0`
+
+If you also want to include the task name, but omit the taskmanager information you can specify the following format:
+
+`metrics.scope.tm.operator: \<host>.\<job_name>.\<task_name>.\<operator_name>.\<subtask_index>`
+
+This could create the name `localhost.MyJob.MySource_->_MyOperator.MyOperator.0`.
+
+The following is a list of all variables available:
```
Comment (uce) [inline docs/apis/common/index.md]: I think it makes sense to make this a subsection so that we can directly link to it etc.

File: docs/apis/common/index.md [hunk 48]
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+    this.histogram.update(value);
+    ...
+  }
+}
+{% endhighlight %}
+
+Flink only provides an interface for Histograms, but offers a Wrapper that allows usage of Codahale/DropWizard Histograms. (org.apache.flink.dropwizard.metrics.DropWizardHistogramWrapper)
+This wrapper is contained in the `flink-metrics-dropwizard` module.
+
+### Scope
```
Comment (uce) [inline docs/apis/common/index.md]: I was wondering whether the names are case sensitive or not?

**Comment by uce** [comment]
Good job Chesnay! Reads very well. I like the way the text and code examples are mixed. I had some minor inline comments.

My biggest concern (not reflected in the inline comments) is the `Scope` section, which is too dense in my opinion. Do you think we can improve it somehow? 

Maybe add sub sections `System Scope`, `User Scope`, `Examples`, `Variables` for starters. The `hierarchical orders` stuff I did not get.

File: docs/apis/common/index.md (hunk 43 above)
Comment (zentol) [inline docs/apis/common/index.md]: JMX for example is not scheduled. I could also imagine Reporters that are queried from the outside.

File: docs/apis/common/index.md [hunk 49]
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
...
+Dependency:
+{% highlight xml %}
+<dependency>
+      <groupId>org.apache.flink</groupId>
+      <artifactId>flink-metrics-ganglia</artifactId>
+      <version>1.1-SNAPSHOT</version>
+</dependency>
+{% endhighlight %}
+
+Parameters:
```
Comment (zentol) [inline docs/apis/common/index.md]: yes

File: docs/apis/common/index.md (hunk 48 above)
Comment (zentol) [inline docs/apis/common/index.md]: the variable names? yes

File: docs/apis/common/index.md (hunk 28 above)
Comment (zentol) [inline docs/apis/common/index.md]: oh, in this case we don't need them since they are in backticks.

File: docs/apis/common/index.md (hunk 31 above)
Comment (zentol) [inline docs/apis/common/index.md]: It's also quite big and blows up the Basic API page. It takes up a quarter of the page alone.

I'll move it.

**Comment by zentol** [comment]
I've made the following changes:
- rewritte scope section
- removed ... from examples
- FIX ALL THE TYPOS
- better line breaking
- added dependency&example for DWHWrapper
- replaced hard-coded dependency versions

File: docs/apis/metrics.md [hunk 50]
```diff
@@ -0,0 +1,441 @@
...
+or a port range, with the default being 9010-9025. The used port is shown in the relevant job or task manager log.
+
+### Ganglia (org.apache.flink.metrics.ganglia.GangliaReporter)
+Dependency:
+{% highlight xml %}
+<dependency>
+      <groupId>org.apache.flink</groupId>
+      <artifactId>flink-metrics-ganglia</artifactId>
+      <version>{{site.version}}</version>
+</dependency>
```
Comment (tillrohrmann) [inline docs/apis/metrics.md]: I think it is not enough to simply include this dependency to your job in order to use the Ganglia reporter. Instead it should be put in the lib directory, because the jar is required when the cluster is started. It only works if the cluster is started for a single job and the user jars are contained in the system classloader.

I think we should document this properly.

Comment (zentol) [inline docs/apis/metrics.md]: can we fix this somehow? I would prefer it if users were not **required** to juggle around jars.

Comment (zentol) [inline docs/apis/metrics.md]: Note: we probably need this long-term anyway (for per-job reporters)

File: docs/apis/metrics.md [hunk 51]
```diff
@@ -0,0 +1,441 @@
...
+- `addressingMode` - UDP addressing mode to use (UNICAST/MULTICAST)
+
+### Graphite (org.apache.flink.metrics.graphite.GraphiteReporter)
+Dependency:
+{% highlight xml %}
+<dependency>
+      <groupId>org.apache.flink</groupId>
+      <artifactId>flink-metrics-graphite</artifactId>
+      <version>{{site.version}}</version>
+</dependency>
```
Comment (tillrohrmann) [inline docs/apis/metrics.md]: We should document that the user has to put the `flink-metrics-graphite-VERSION-jar-with-dependencies.jar` in the `lib` folder or make it differently accessible to the cluster when starting it.

Comment (tillrohrmann) [inline docs/apis/metrics.md]: The same probably applies to the `GangliaReporter`.

File: docs/apis/metrics.md (hunk 50 above)
Comment (tillrohrmann) [inline docs/apis/metrics.md]: At the moment I fear there is no way around this, since we start the reporters when the cluster is started.

Comment (zentol) [inline docs/apis/metrics.md]: right, of course. I've adjusted the docs.

File: docs/apis/metrics.md [hunk 52]
```diff
@@ -0,0 +1,441 @@
...
+<dependency>
+      <groupId>org.apache.flink</groupId>
+      <artifactId>flink-metrics-ganglia</artifactId>
+      <version>{{site.version}}</version>
+</dependency>
+{% endhighlight %}
+
+Parameters:
+
+- `host` - the ganglia server host
```
Comment (tillrohrmann) [inline docs/apis/metrics.md]: We should highlight that this is the value you set in `gmond.conf` `udp_recv_channel.bind`. Otherwise the UPD packages won't arrive. That was actually the problem I had the whole time. I'm just not good with tooling...

Comment (zentol) [inline docs/apis/metrics.md]: Note that it is not mandatory to set the bind property. If it is not set then generally host should be set to localhost.

But yes, if it set it to something but localhost then naturally you have to set it to that value, will include that in the docs.

Comment (tillrohrmann) [inline docs/apis/metrics.md]: The `udp_recv_channel.bind` is set per default in the `gmond.conf` if you set up Ganglia.
